File: src/WallpaperReddit/Services/Logger.cs
using System;
using System.IO;
using System.Text;

namespace WallpaperReddit.Services
{
    /// <summary>Tiny append-only logger with size-based rollover. No dependencies.</summary>
    public static class Logger
    {
        private static readonly object Gate = new();
        private const long MaxBytes = 1_000_000; // ~1 MB then roll once

        public static event Action<string> OnLine;

        public static void Info(string message) => Write("INFO", message);
        public static void Warn(string message) => Write("WARN", message);

        public static void Error(string message, Exception ex = null)
        {
            var text = ex == null ? message : $"{message} :: {ex.GetType().Name}: {ex.Message}";
            Write("ERROR", text);
        }

        private static void Write(string level, string message)
        {
            var line = $"{DateTime.Now:yyyy-MM-dd HH:mm:ss} [{level}] {message}";
            try
            {
                lock (Gate)
                {
                    RollIfNeeded();
                    File.AppendAllText(AppPaths.LogFile, line + Environment.NewLine, Encoding.UTF8);
                }
            }
            catch
            {
                // Logging must never crash the app.
            }

            try { OnLine?.Invoke(line); } catch { /* UI listener issue is non-fatal */ }
        }

        private static void RollIfNeeded()
        {
            try
            {
                var fi = new FileInfo(AppPaths.LogFile);
                if (fi.Exists && fi.Length > MaxBytes)
                {
                    var backup = AppPaths.LogFile + ".1";
                    if (File.Exists(backup)) File.Delete(backup);
                    File.Move(AppPaths.LogFile, backup);
                }
            }
            catch { /* ignore */ }
        }
    }
}

File: kde/rwr/logger.py
"""Tiny logger: appends to the log file and fans out lines to UI listeners."""
from __future__ import annotations

import datetime
from typing import Callable, List

from . import paths

_listeners: List[Callable[[str], None]] = []


def add_listener(fn: Callable[[str], None]) -> None:
    _listeners.append(fn)


def remove_listener(fn: Callable[[str], None]) -> None:
    if fn in _listeners:
        _listeners.remove(fn)


def _write(level: str, message: str) -> None:
    line = f"{datetime.datetime.now():%Y-%m-%d %H:%M:%S} [{level}] {message}"
    try:
        paths.LOG_FILE.parent.mkdir(parents=True, exist_ok=True)
        # Roll once at ~1 MB.
        if paths.LOG_FILE.exists() and paths.LOG_FILE.stat().st_size > 1_000_000:
            paths.LOG_FILE.replace(paths.LOG_FILE.with_suffix(".log.1"))
        with paths.LOG_FILE.open("a", encoding="utf-8") as fh:
            fh.write(line + "\n")
    except Exception:
        pass  # logging must never crash the app
    for fn in list(_listeners):
        try:
            fn(line)
        except Exception:
            pass


def info(msg: str) -> None:
    _write("INFO", msg)


def warn(msg: str) -> None:
    _write("WARN", msg)


def error(msg: str) -> None:
    _write("ERROR", msg)
